shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,773 @@
+  constructor(array) {
+    if (!array) {
+      this.array = [];
+    } else if (array instanceof Array) {
+      this.array = array.slice(0);
+    } else {
+      throw new Error("don't know how to interpret as PermissionSet: " + array);
+    }
   }
-}
 
-// Methods for mutating a PermissionSet by combining it with another PermissionSet.
-// These return a boolean indicating whether the operation had any effect.
+  static fromRoleAssignment(roleAssignment, viewInfo) {
+    let result = new PermissionSet([]);
+
+    if (!roleAssignment || "none" in roleAssignment) {
+      if (viewInfo.roles) {
+        for (let ii = 0; ii < viewInfo.roles.length; ++ii) {
+          const roleDef = viewInfo.roles[ii];
+          if (roleDef.default) {
+            result = new PermissionSet(roleDef.permissions);
+            break;
+          }
+        }
+      }
+    } else if ("allAccess" in roleAssignment) {
+      let length = 0;
+      if (viewInfo.permissions) {
+        length = viewInfo.permissions.length;
+      }
 
-PermissionSet.prototype.add = function(other) {
-  var changed = false;
-  for (var ii = 0; ii < other.array.length; ++ii) {
-    var old = this.array[ii];
-    this.array[ii] = this.array[ii] || other.array[ii];
-    if (old != this.array[ii]) {
-      changed = true;
+      const array = new Array(length);
+      for (let ii = 0; ii < array.length; ++ii) {
+        array[ii] = true;
+      }
+
+      result = new PermissionSet(array);
+    } else if ("roleId" in roleAssignment && viewInfo.roles && viewInfo.roles.length > 0) {
+      const roleDef = viewInfo.roles[roleAssignment.roleId];
+      if (roleDef) {
+        result = new PermissionSet(roleDef.permissions);
+      }
     }
+
+    if (roleAssignment) {
+      result.add(new PermissionSet(roleAssignment.addPermissionSet));
+      result.remove(new PermissionSet(roleAssignment.removePermissionSet));
+    }
+
+    return result;
   }
-  return changed;
-}
 
-PermissionSet.prototype.remove = function(other) {
-  var changed = false;
-  for (var ii = 0; ii < other.array.length && ii < this.array.length; ++ii) {
-    var old = this.array[ii];
-    this.array[ii] = this.array[ii] && !other.array[ii];
-    if (old != this.array[ii]) {
-      changed = true;
+  isSubsetOf(other) {
+    check(other, PermissionSet);
+    for (let ii = 0; ii < this.array.length; ++ii) {
+      const mine = this.array[ii];
+      const yours = other.array[ii] || false;
+      if (mine != yours) {
+        return false;
+      }
     }
+
+    return true;
+  }
+
+  // Methods for mutating a PermissionSet by combining it with another PermissionSet.
+  // These return a boolean indicating whether the operation had any effect.
+
+  add(other) {
+    check(other, PermissionSet);
+    let changed = false;
+    for (let ii = 0; ii < other.array.length; ++ii) {
+      const old = this.array[ii];
+      this.array[ii] = this.array[ii] || other.array[ii];
+      if (old != this.array[ii]) {
```

Comment:
If `old` is `undefined` but the new value is `false`, then the PermissionSet hasn't really changed, but this will treat it as having changed.

Also, please use `!==` rather than `!=`.